control-plane/subcommand/sync-catalog/command.go

```diff
@@ -393,6 +415,85 @@ func (c *Command) Run(args []string) int {
 	}
 }
 
+// remove all k8s services from Consul.
+func (c *Command) removeAllK8SServicesFromConsulNode(consulClient *api.Client, nodeName string) error {
+	node, _, err := consulClient.Catalog().NodeServiceList(nodeName, &api.QueryOptions{Filter: c.flagFilter})
+	if err != nil {
+		return err
+	}
```

Comment:
It might be more efficient to use [`Catalog().Services()`](https://pkg.go.dev/github.com/hashicorp/consul/api#Catalog.Services) to obtain the list of services when no user-defined filter is provided. Pass in a filter such as `Node=="<nodeName>"` to return a list of services that exist on that node. Count the number of logical services on that node and prompt for user confirmation before deleting the node. This will purge all of the instances associated with the previously returned logical services that exist on that node.

If the user provides their own filter, use `Catalog().NodeServiceList()` query the list of service *instances* that are registered against that node and that match the specified filter criteria.

Reply:
Thanks for review! @blake 
I checked the API for [listing services](https://developer.hashicorp.com/consul/api-docs/catalog#list-services), it looks like the response is the structure of [map[string][]string](https://github.com/hashicorp/consul/blob/main/api/catalog.go#L212) whose keys are [service names](https://developer.hashicorp.com/consul/api-docs/catalog#sample-response-2). However, in our de-registration request, we need to have the information of [serviceID](https://developer.hashicorp.com/consul/api-docs/catalog#serviceid). So I think in our situation, [Catalog().NodeServiceList()](https://developer.hashicorp.com/consul/api-docs/catalog#list-services-for-node) suites better to our needs. Please point out something I overlooked here, thank you!

Reply:
You don't need the service ID when de-registering a node. You need to provide the node ID. All services on the node will then be removed. <https://developer.hashicorp.com/consul/api-docs/catalog#node-1>

> Specifies the ID of the node. If no other values are provided, this node, all its services, and all its checks are removed.